drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2018 Philémon Jaermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <i2c.h>
+#include <init.h>
+#include <sensor.h>
+#include <logging/sys_log.h>
+
+#include "lsm303dlhc_magneto.h"
+
+static int lsm303dlhc_sample_fetch(struct device *dev,
+		enum sensor_channel chan)
+{
+	struct lsm303dlhc_magneto_data *drv_data = dev->driver_data;
+	u8_t tries = 11;
+	u8_t magneto_buf[6];
+	u8_t status;
+
+	/* Check data ready flag */
+	while (tries-- > 0) {
+		if (i2c_reg_read_byte(drv_data->i2c,
+					LSM303DLHC_MAGN_I2C_ADDR,
+					LSM303DLHC_SR_REG_M,
+					&status) < 0) {
+			SYS_LOG_ERR("Failed to read status register.");
+			return -EIO;
+		}
+
+		if ((status & LSM303DLHC_MAGN_DRDY) || tries == 0) {
+			break;
+		}
+
+		k_sleep(100);
```

Comment:
I think that a fixed sleep to 100ms does not work.
If you have set the odr to <= 1hz you may always find DRDY to 0.
If you have set the odr to high freq you wait too long.

I think that a variable time based on odr would be suit better.

Reply:
I agree for the variable time

Reply:
ok, and if you do in this case 10 retries (10 ODRs basically) are too much now.
You may want to decrease to 3/4. What do you think?

Reply:
Mh I thought about keeping the 10 retries but having the sleep time to be ODR/10, or maybe even 11 retries if the querry is perfectly synchronised with the ODR.